src/main/java/com/github/junrar/rarfile/FileHeader.java

```diff
@@ -186,9 +192,71 @@ public FileHeader(BlockHeader bh, byte[] fileHeader) {
                 position++;
             }
         }
-        mTime = getDateDos(fileTime);
-        // TODO rartime -> extended
 
+        mTime = FileTime.fromMillis(getDateDos(fileTime));
+
+        if (hasExtTime()) {
+            short extTimeFlags = Raw.readShortLittleEndian(fileHeader, position);
+            position += 2;
+
+            TimePositionTuple mTimeTuple = parseExtTime(12, extTimeFlags, fileHeader, position, mTime);
+            mTime = mTimeTuple.time;
+            position = mTimeTuple.position;
+
+            TimePositionTuple cTimeTuple = parseExtTime(8, extTimeFlags, fileHeader, position);
+            cTime = cTimeTuple.time;
+            position = cTimeTuple.position;
+
+            TimePositionTuple aTimeTuple = parseExtTime(4, extTimeFlags, fileHeader, position);
+            aTime = aTimeTuple.time;
+            position = aTimeTuple.position;
+
+            TimePositionTuple arcTimeTuple = parseExtTime(0, extTimeFlags, fileHeader, position);
+            arcTime = arcTimeTuple.time;
+            position = arcTimeTuple.position;
+        }
+    }
+
+    private static final class TimePositionTuple {
+        private final int position;
+        private final FileTime time;
+
+        private TimePositionTuple(int position, FileTime time) {
+            this.position = position;
+            this.time = time;
+        }
+    }
+
+    private static TimePositionTuple parseExtTime(int shift, short flags, byte[] fileHeader, int position) {
+        return parseExtTime(shift, flags, fileHeader, position, null);
+    }
+
+    private static TimePositionTuple parseExtTime(int shift, short flags, byte[] fileHeader, int position, FileTime baseTime) {
+        int flag = flags >>> shift;
+        if ((flag & 0x8) != 0) {
+            long seconds;
+            if (baseTime != null) {
+                seconds = baseTime.toMillis() / MILLIS_PER_SECOND;
+            } else {
+                seconds = getDateDos(Raw.readIntLittleEndian(fileHeader, position)) / MILLIS_PER_SECOND;
+                position += 4;
+            }
+            int count = flag & 0x3;
+            long remainder = 0;
+            for (int i = 0; i < count; i++) {
+                int b = fileHeader[position] & 0xff;
+                remainder = (b << 16) | (remainder >>> 8);
+                position++;
+            }
+            long nanos = remainder * NANOS_PER_UNIT;
+            if ((flag & 0x4) != 0) {
```

Comment:
Are you sure about this one? It should be the second bit, and this is seems to check the 4th instead, no? I'm not very fluent in bitwise operation, but this also differs from the pseudo code from the website you referenced, where they use `4` and not `0x4`.

Reply:
4 and 0x4 are the same thing.

4 * 10^0 = 4 * 16^0 = 1*2^2 + 0*2^1 + 0* 2^0

So 4 here is 0100, which is the second bit.

Reply:
BTW do you think you could add a test for an archive that needs the 1s correction?

Reply:
That branch is covered in the archives I used for testing, isn't it?

The archive I added for testing needs this correction. Without the correction its modification time ends in 54 seconds. With it, it ends in 55, and that's checked in the tests.

Or do you mean a specific test for that?

Reply:
my bad, i didn't check the coverage 🤦🏻